prog/dftb+/lib_dftb/sparse2dense.F90

```diff
@@ -1559,4 +1425,952 @@ subroutine blockAntiSymmetrizeHS_real(square, iAtomStart)
 
   end subroutine blockAntiSymmetrizeHS_real
 
+
+#:if WITH_SCALAPACK
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!! Scalapack routines
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+
+  !> Unpacks sparse H or S into dense (real, blacs).
+  !>
+  !> Note: In contrast to the serial routines, both triangles of the resulting matrix are filled.
+  !>
+  subroutine unpackHSRealBlacs(myBlacs, orig, iNeighbor, nNeighbor, iPair, img2CentCell, desc,&
+      & square)
+
+    !> BLACS matrix descriptor
+    type(TBlacsEnv), intent(in) :: myBlacs
+
+    !> sparse matrix
+    real(dp), intent(in) :: orig(:)
+
+    !> Neighbor list for each atom (First index from 0!)
+    integer, intent(in) :: iNeighbor(0:,:)
+
+    !> Nr. of neighbors for each atom (incl. itself).
+    integer, intent(in) :: nNeighbor(:)
+
+    !> indexing array for the sparse Hamiltonian
+    integer, intent(in) :: iPair(0:,:)
+
+    !> Map from images of atoms to central cell atoms
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Dense matrix description
+    type(TDenseDescr), intent(in) :: desc
+
+    !> dense matrix part of distributed whole
+    real(dp), intent(out) :: square(:,:)
+
+    integer :: nAtom
+    integer :: iOrig, ii, jj, nOrb1, nOrb2
+    integer :: iNeigh
+    integer :: iAtom1, iAtom2, iAtom2f
+
+    nAtom = size(iNeighbor, dim=2)
+
+    @:ASSERT(nAtom > 0)
+    @:ASSERT(size(nNeighbor) == nAtom)
+    @:ASSERT(size(desc%iAtomStart) == nAtom + 1)
+
+    square(:,:) = 0.0_dp
+
+    do iAtom1 = 1, nAtom
```

Comment:
Might be worth adding a comment that this is linear scaling in system size, so does not incur a large cost for running on all processes in the group.